src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/vnode_fusion.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "vnode_fusion.hpp"
+
+#include <cstdint>
+#include <limits>
+
+#include "itt.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/opsets/opset1.hpp"
+#include "openvino/opsets/opset6.hpp"
+#include "openvino/opsets/opset8.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
+#include "openvino/pass/pattern/op/or.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "ov_ops/type_relaxed.hpp"
+#include "transformations/cpu_opset/common/op/vnode.hpp"
+#include "transformations/utils/utils.hpp"
+#include "utils/gen_pattern.hpp"
+
+using namespace ov::gen_pattern;
+
+class CausalMaskPreprocess : public ov::pass::MatcherPass {
+public:
+    OPENVINO_RTTI("CausalMaskPreprocess", "0");
+    CausalMaskPreprocess();
+};
+
+/*
+following pattern is from:
+    LlamaModel._update_causal_mask() models/llama/modeling_llama.py:
+    GemmaModel._update_causal_mask() models/llama/modeling_gemma.py:
+
+the python code is:
+
+    min_dtype = torch.finfo(dtype).min
+    causal_mask = self.causal_mask[None, None, :, :].repeat(batch_size, 1, 1, 1).to(dtype) * min_dtype
+    causal_mask = causal_mask.to(dtype=dtype, device=device)
+
+    mask_length = attention_mask.shape[-1]
+    padding_mask = causal_mask[..., :mask_length].eq(0.0) * attention_mask[:, None, None, :].eq(0.0)
+    causal_mask[..., :mask_length] = causal_mask[..., :mask_length].masked_fill(padding_mask, min_dtype)
+
+    ######## when being used will be further sliced
+    causal_mask = attention_mask
+        if attention_mask is not None and cache_position is not None:
+            causal_mask = causal_mask[:, :, cache_position, : key_states.shape[-2]]
+
+inputs:
+    causal mask (slice_Slice) : boolean[1, 1, maxL, maxL]
+    attention_mask            : i64[N, kv_len]
+                                 0 means mask-out, 1 means attends to
+    batch_size (size_Gather)  : i32[1]
+    cache_positions  i32[q_len];
+    kvLen            i32[1];
+
+outputs:
+    causal mask for SDPA : f32[batch_size, 1, q_len, kvLen]
+*/
+
+template <typename T>
+bool is_triu(ov::opset1::Constant* cmask, size_t rows, size_t columns) {
+    const auto* ptr = reinterpret_cast<const T*>(cmask->get_data_ptr());
+    for (size_t y = 0; y < rows; y++, ptr += columns) {
+        size_t x;
+        for (x = 0; x <= y; x++)
+            if (ptr[x])
+                return false;
+        for (; x < columns; x++)
+            if (!ptr[x])
+                return false;
+    }
+    return true;
+}
+
+CausalMaskPreprocess::CausalMaskPreprocess() {
```

Comment:
Test case is needed to test this fusion,

Reply:
OK